### ISSUE TITLE
Sprint Week: Oct 22-26 2018 - Royal Observatory Edinburgh

### ISSUE THREAD

**Issue by rbliu** [rbliu/jedisim#12]
During this Sprint Week, we plan to

- [x] Debug and troubleshoot the `jedidistort` code.

- [x] Code review and improve memory usage.

- [x] Run validation tests with deflection angle map of known NFW-profile clusters.

**Comment by rbliu** [comment]
**Day 1**

- [x] Initial commit made. 

- [x] Working on issue:
```
$ ./jedidistortDC2 4096 4096 trial_6/trial6_dislist.txt physics_settings/production_1_clusters/lenses_10.000000_4.000000.txt 0.03 0.3
Running jedidistort on image of shape (4096, 4096).
Pixel scale: 0.030000.
Lens redshift: 0.300000.

Opened galaxy list file "trial_6/trial6_dislist.txt".
1000 galaxies in "trial_6/trial6_dislist.txt".
Opened lens list file "physics_settings/production_1_clusters/lenses_10.000000_4.000000.txt".
1 lenses in "physics_settings/production_1_clusters/lenses_10.000000_4.000000.txt".
Lens 0:
x: 2048.000000
y: 2048.000000
type: 2
p1: 10.000000
p2: 4.000000

dx: 2048.000000
dy: 2048.000000
dr: 2896.309326
nr: 28964

Dl: 946.326477
rs: 0.473513
rdelta_c: 1.235425
delta_c: 5271.147664
NFW_prefactor: 1.716197e+00

Calculated profile for lens 0.

Read in alpha file1 successfully! 
Read in alpha file2 successfully! 
alpha arr1 memory allocated successfully! 
alpha arr2 memory allocated successfully! 
making grid 0 with sieve 8: (512,512).
Segmentation fault: 11
```

**Comment by rbliu** [comment]
**Day 2**

- [x] Generated an initial sample image!
(4096x4096pix @0.03''/pix ==> 470x470pix @0.2''/pix)
Left: simulation image. Right: the image overlapped with the truth kappa map.
![image](https://user-images.githubusercontent.com/10132464/47388975-87417600-d6e1-11e8-8572-63cdba01c80e.png)

- [x] A potential issue: 
The seg fault appears occasionally during the distorting step. I will look into it...
_Update_: In a sample simulation where this segmentation fault happened, I modified the `dislist.txt` by moving the stuck entry forward -- it works! So I guess the image/catalog/alpha-map should be good, but sometimes the code might cause memory error on my laptop. 

- [ ] Improve memory usage:
Now the alpha map files are first converted into text format and fed into `jedisim`. To reduce the memory occupied, I plan to keep their `.bin` format and open directly in the code.

**Comment by rbliu** [comment]
**Day 3 + 4**

- [x] Make an NFW alpha map.
A pixel-by-pixel deflection angle map is calculated based on an NFW-profile cluster (M200=1E15Msun, concentration=4.0). 

- [x] Validation test:
Compare the simulation images lensed by an analytical NFW profile and the numerical alpha map generated above. 
Two examples are shown below:
![image](https://user-images.githubusercontent.com/10132464/47524548-e1207800-d868-11e8-9950-677ab3c35b83.png)
![image](https://user-images.githubusercontent.com/10132464/47524552-e4b3ff00-d868-11e8-87a7-a5a64610cc32.png)
_Left_: simulation with an analytical NFW profile.
_Middle_: simulation (same galaxy sample) with numerical NFW alpha map, processed by the new pipeline.
_Right_: residual between these two images. 
Residual in galaxy area ~0, while it fluctuates in the background area due to random noise added when generating the simulations.

**Comment by rbliu** [comment]
**Day 5**

- [x] Minor cleaning on the code, and wrap-up! @linan7788626

**Comment by linan7788626** [comment]
> **Day 3 + 4**
> 
> * [x]  Make an NFW alpha map.
>   A pixel-by-pixel deflection angle map is calculated based on an NFW-profile cluster (M200=1E15Msun, concentration=4.0).

A quick question about this, you have an NFW halo, then you calculate the deflection angles by using the analytic NFW-deflection model (alpha_vector_0) and FFT-numerical method (alpha_vector_1). The validation test is: comparing the lensed images with alpha_vector_0 and alpha_vector_1 respectively. Do I understand correctly?  @rbliu

**Comment by rbliu** [comment]
> > **Day 3 + 4**
> > 
> > * [x]   Make an NFW alpha map.
> >   A pixel-by-pixel deflection angle map is calculated based on an NFW-profile cluster (M200=1E15Msun, concentration=4.0).
> 
> A quick question about this, you have an NFW halo, then you calculate the deflection angles by using the analytic NFW-deflection model (alpha_vector_0) and FFT-numerical method (alpha_vector_1). The validation test is: comparing the lensed images with alpha_vector_0 and alpha_vector_1 respectively. Do I understand correctly? @rbliu

@linan7788626 Correct. This is a test on the updated code -- we are expecting to generate the same output image with either (alpha_vector_0) or (alpha_vector_1), which means the code is distorting galaxies correctly. Therefore, we can trust its performance on DC2 alpha maps.

**Comment by linan7788626** [comment]
@rbliu thanks for the updates, Just wrapped-up for you here, nice work. I will open a new issue about the suggestions for the next step. Cheers.
